Client hooks

devintang3 opened the pull request:
Continuation of https://github.com/jupyter/nbclient/pull/79

I rebased with master and added some tests. I occasionally get errors on `test_many_parallel_notebooks` which I don't think is related, but I'll also get it on `test_error_cell_hooks` in py38. It's an intermittent error that gets resolved if I add a sleep, so I think it's something related to the async task not running right away. Any guidance here would be appreciated.

I also removed py36 from tox since it looked like the minimum version is now Python3.7 and it was just hanging when I tried running it.

davidbrochart commented on nbclient/util.py:
```diff
@@ -102,3 +103,15 @@ async def ensure_async(obj: Union[Awaitable, Any]) -> Any:
+        loop = asyncio.get_event_loop()
+        hook_with_kwargs = partial(hook, **kwargs)
+        future = loop.run_in_executor(None, hook_with_kwargs)
+    asyncio.ensure_future(future)
```
I don't understand, does it mean that hooks are going to be launched in the background? BTW, it would be nice to test async hooks.
I think `run_hook` should be `async`:
```python
async def run_hook(hook: Optional[Callable], **kwargs) -> None:
    await ensure_async(hook(**kwargs))
```

devintang3 replied:
Yeah, the intent is for the hook to be launched in the background.

davidbrochart reviewed (changes requested):
Thanks a lot for this PR @devintang3.
Would you like to add these new features to the documentation also?

devintang3 commented:
Thanks for the suggestions. I've applied your recommendations as well as added more tests. 
Some changes to note:
* tests/util.py has been renamed to tests/test_util.py so that pytest can pick it up automatically
* Added the package pytest-asyncio to requirements-dev.txt so I can test some async methods
* Updated docs
* The pre-commit hook was complaining about unused variables so I've removed them from test_client.py. I'm not quite sure what the purpose of the `loop = asyncio.get_event_loop()` was

davidbrochart commented on nbclient/client.py:
```diff
@@ -245,6 +245,52 @@ class NotebookClient(LoggingConfigurable):
+        default_value=None,
+        allow_none=True,
+        help=dedent(
+            """
+        Called after the kernel manager and kernel client are setup, and cells
+        are about to execute.
+        """
```
You didn't make the change.

devintang3 replied:
Thanks! Not sure how I missed this one.

davidbrochart commented on nbclient/util.py:
```diff
@@ -102,3 +103,15 @@ async def ensure_async(obj: Union[Awaitable, Any]) -> Any:
+async def run_hook(hook: Optional[Callable], **kwargs) -> None:
+    if hook is None:
+        return
+    if asyncio.iscoroutinefunction(hook):
+        future = hook(**kwargs)
+    else:
+        loop = asyncio.get_event_loop()
+        hook_with_kwargs = partial(hook, **kwargs)
+        future = loop.run_in_executor(None, hook_with_kwargs)
+    await asyncio.ensure_future(future)
```
```suggestion
    res = hook(**kwargs)
    if inspect.isawaitable(res):
        await res
```

davidbrochart commented:
Could you rebase on master? It looks like you merged master into your branch instead.

chrisjsewell commented on nbclient/client.py:
```diff
@@ -805,11 +865,13 @@ async def async_execute_cell(
             self.allow_errors or "raises-exception" in cell.metadata.get("tags", [])
         )
 
+        await run_hook(self.on_cell_start, cell=cell, cell_index=cell_index)
```
while we are at it, perhaps there could be a hook at the very start of this function, before even skipping non-executing cells, like `on_cell_visit`

devintang3 replied:
Does it make sense to just move `on_cell_start` to the start instead of creating a new hook?

davidbrochart replied:
That would make sense, and we know if the cell is a Markdown cell with `cell.cell_type`.
But then maybe add `on_cell_execute` right before the execution? It won't be called if the cell is not a code cell or if the cell is skipped because of a tag.

devintang3 replied:
I moved `on_cell_start` towards the top so it'll execute for non-code cells and added  a new hook called `on_cell_execute`

choldgraf commented on nbclient/client.py:
```diff
@@ -426,6 +483,7 @@ async def async_start_new_kernel_client(self) -> KernelClient:
             await self._async_cleanup_kernel()
             raise
         self.kc.allow_stdin = False
+        await run_hook(self.on_execution_start)
```
```suggestion
        await run_hook(self.on_execution_start, notebook=self.nd)
```

What about something like this so that people have access to the NotebookNode before execution?

choldgraf commented on nbclient/client.py:
```diff
@@ -245,6 +256,52 @@ class NotebookClient(LoggingConfigurable):
 
     kernel_manager_class: KernelManager = Type(config=True, help='The kernel manager class to use.')
 
+    on_execution_start: t.Optional[t.Callable] = Callable(
```
I think `on_notebook_start` might be more intuitive. `on_execution_start` makes me think that it will be called as soon as the execution begins, whereas this is actually called when the notebook is initialized but no cells have been identified for execution yet.

choldgraf commented on nbclient/client.py:
```diff
@@ -426,6 +483,7 @@ async def async_start_new_kernel_client(self) -> KernelClient:
             await self._async_cleanup_kernel()
             raise
         self.kc.allow_stdin = False
+        await run_hook(self.on_execution_start)
```
For symmetry's sake, why not also include two extra notebook-level hooks: `on_notebook_complete` and `on_notebook_error`, that map on to the two cell execution hooks but at a notebook level?

That way you would have the same basic pattern for both the notebook, and for each cell in the notebook:

- hook for just before execution happens (`on_notebook_start`, `on_cell_start`)
- hook for when execution completes (`on_notebook_complete`, `on_cell_complete`)
- hook for when execution has an error (`on_notebook_error`, `on_cell_error`)

devintang3 replied:
Added `on_notebook_error`, although I'm not sure if I did that part correctly. The only time I've personally seen the notebook fail is when there's a RuntimeError, so that's where I've placed the hook.

choldgraf replied:
fair enough - and feel free to push back if you think that this is "too many hooks all at once" 😅 on the one hand I like symmetry and intentional design for extension points. On the other hand, I also believe in not building things until somebody actively wants it to be built :-)  so if you think this is too much complexity, another option could be to just add comments for where you think the extra hooks could go, and wait to build them until a user explicitly asks for it.

choldgraf reviewed (commented):
I think this is a really nice extension for `nbclient` - I have a few comments and suggestions, mostly around the API design and the documentation to make it clear how to use it

davidbrochart commented on docs/client.rst:
```diff
@@ -96,6 +96,15 @@ on both versions. Here the traitlet ``kernel_name`` helps simplify and
 maintain consistency: we can just run a notebook twice, specifying first
 "python2" and then "python3" as the kernel name.
 
+In addition to the two above, we also support traitlets for hooks. They are as
+follows: ``on_execution_start``, ``on_cell_start``, ``on_cell_complete``,
+``on_cell_error``. These traitlets allow specifying a ``Callable`` function,
+which will run at certain points during the notebook execution and is executed asynchronously.
+``on_execution_start`` will run when the notebook client is kicked off.
+``on_cell_start`` will run right before each cell is executed.
+``on_cell_complete`` will run right after the cell is executed.
+``on_cell_error`` will run if there is an error in the cell.
+
```
Maybe worth mentioning that the hooks *can* be async, but don't have to be?

devintang3 replied:
Thanks @choldgraf! Much appreciated.

I removed the line about functions can be async. If it can be async or sync, then I don't see much point in mentioning that

davidbrochart commented on docs/client.rst:
```diff
@@ -96,6 +96,36 @@ on both versions. Here the traitlet ``kernel_name`` helps simplify and
 maintain consistency: we can just run a notebook twice, specifying first
 "python2" and then "python3" as the kernel name.
 
+Hooks before and after cell execution
+-------------------------------------
```
```suggestion
Hooks before and after notebook or cell execution
-------------------------------------------------
```

davidbrochart commented on docs/client.rst:
```diff
@@ -96,6 +96,36 @@ on both versions. Here the traitlet ``kernel_name`` helps simplify and
+Hooks before and after cell execution
+-------------------------------------
+There are several configurable hooks that allow the user to execute code before and
+after a cell is executed. Each one is configured with a function that will be called in its
+respective place in the cell execution pipeline.
```
```suggestion
after a notebook or a cell is executed. Each one is configured with a function that will be called in its
respective place in the execution pipeline.
```

davidbrochart commented on docs/client.rst:
```diff
@@ -96,6 +96,36 @@ on both versions. Here the traitlet ``kernel_name`` helps simplify and
+**Cell-level hooks**: These hooks are called with two parameters:
+
+- ``cell=NotebookNode``: a reference to the current cell.
+- ``cell_index=int``: the index of the cell in the current notebook's list of cells
```
```suggestion
- ``cell_index=int``: the index of the cell in the current notebook's list of cells.
```

davidbrochart commented on nbclient/client.py:
```diff
@@ -261,6 +272,87 @@ class NotebookClient(LoggingConfigurable):
+        help=dedent(
+            """
+            Called when the notebook encountered an error.
+            Called with kwargs `notebook`
```
```suggestion
            Called with kwargs `notebook`.
```

davidbrochart commented on nbclient/client.py:
```diff
@@ -261,6 +272,87 @@ class NotebookClient(LoggingConfigurable):
+        help=dedent(
+            """
+            Called after kernel is cleaned up.
+            Called with kwargs `notebook`
```
```suggestion
            Called with kwargs `notebook`.
```

davidbrochart commented on nbclient/client.py:
```diff
@@ -261,6 +272,87 @@ class NotebookClient(LoggingConfigurable):
+            """
+            Called after the kernel manager and kernel client are setup, and cells
+            are about to execute.
+            Called with kwargs `notebook`
```
```suggestion
            Called with kwargs `notebook`.
```

davidbrochart commented on nbclient/client.py:
```diff
@@ -513,10 +607,13 @@ def on_signal():
             await self.async_start_new_kernel_client()
         try:
             yield
+        except RuntimeError:
+            await run_hook(self.on_notebook_error, notebook=self.nb)
+            raise
```
```suggestion
        except RuntimeError as e:
            await run_hook(self.on_notebook_error, notebook=self.nb)
            raise e
```

davidbrochart commented on nbclient/tests/test_client.py:
```diff
@@ -345,11 +345,7 @@ def test_async_parallel_notebooks(capfd, tmpdir):
-        ]
-        loop = asyncio.get_event_loop()
-        loop.run_until_complete(asyncio.gather(*tasks))
+        [async_run_notebook(input_file.format(label=label), opts, res) for label in ("A", "B")]
```
You removed the execution of these tasks, can you explain why?

devintang3 replied:
I think this was a mistake on my part when I was trying to investigate some failures. I've added those changes back in.

davidbrochart commented on nbclient/tests/test_client.py:
```diff
@@ -370,9 +366,7 @@ def test_many_async_parallel_notebooks(capfd):
-    tasks = [async_run_notebook(input_file, opts, res) for i in range(4)]
-    loop = asyncio.get_event_loop()
-    loop.run_until_complete(asyncio.gather(*tasks))
+    [async_run_notebook(input_file, opts, res) for i in range(4)]
```
Same here.

davidbrochart reviewed (approved):
Thanks a lot @devintang3, that looks good to me. A couple of notes, but I think this PR can be merged anyway:
- maybe use in the tests: `hooks  = [MagicMock() for i in range(7)]`.
- have all the hook help strings start with `A callable which executes...`.

devintang3 commented:
> * maybe use in the tests: `hooks  = [MagicMock() for i in range(7)]`.
> * have all the hook help strings start with `A callable which executes...`.

That makes sense to me. I've updated the tests to do a loop and updated the help strings as well

davidbrochart commented:
Thanks @devintang3 !

choldgraf commented:
🚀
